### DISCUSSION TITLE
Is there a nicer way to communicate between components on different levels (or accesing the application window deep below)?

### DISCUSSION THREAD

**Discussion by paveloom** [Relm4/Relm4#212] [Q&A] [answered]
So, imagine you have the following graph of components:

```mermaid
flowchart LR
Window --> A --> B --> C
Window --> Dialog
```

The dialog is set to be modal and transient for the parent window, so it can't be moved. However, its sole purpose is to communicate with the `C` component, so it would naturally belong as a component of the `C` component. There are basically two diametrical problems here:

1) If the graph is left as is, one should construct a channel between the two components by creating messages in *each* of the components in between: this pollutes the code of the components and leads to quite a lot of duplication.
2) If the `Dialog` component is moved to be a component of the `C` component, there is no way to make it transient for the parent window. Well, one could copy the parent window into each component (similar to the [`save_dialog`](https://github.com/Relm4/Relm4/blob/8985b76ae0ac612b63c0830fb8c6ec85327c3af2/relm4-examples/examples/save_dialog.rs) example, but saving the window into an additional field of *each* component in between), but that's just a waste of memory (also, pollutes the model of each component).

Are there nicer ways to handle either of these problems?

**Comment by AaronErhardt** [comment, marked as answer]
So far, there isn't a simpler way to handle this but I transferred this to #213 which is part of the 0.5 milestone. The question is how we can implement a better solution.

The ideas that come to my mind right now are:

1. Global, lazily initialized components. But then it would be difficult to pass the parent window...
2. A global message broker that passes messages to a certain component.

Maybe I'll have better ideas tomorrow, I'm too tired right now :sleeping:
